pthread_join test hangs occasionally

A CI run of `basic-pthread_join-(debug)-(run-hw)-(8-ethreads)` led to the following hanging behaviour in the middle of the test:
```
Execution timeout: 60
Will run tests for run-hw
/opt/sgx-lkl-debug/bin/sgx-lkl-disk create --size=5M --docker=./Dockerfile sgx-lkl-rootfs.img
Using detected host DNS server in /etc/resolv.conf: 168.63.129.16
Creating sgx-lkl-rootfs.img from Dockerfile ./Dockerfile...
Building Docker image...
Creating disk image file from Docker container...
Succesfully created sgx-lkl-rootfs.img.
SGXLKL_VERBOSE=1 SGXLKL_KERNEL_VERBOSE=1 /opt/sgx-lkl-debug/bin/sgx-lkl-run-oe --hw-debug sgx-lkl-rootfs.img pthread_join-test
[   SGX-LKL  ] nproc=4 ETHREADS=8 CMDLINE="mem=32M"
[   SGX-LKL  ] HW TLS support: econf->fsgsbase=1
[   SGX-LKL  ] Registering disk 0 (path='sgx-lkl-rootfs.img', mnt='/', [RW])
[   SGX-LKL  ] No tap device specified, networking will not be available.
[   SGX-LKL  ] get_signed_libsgxlkl_path... result=/opt/sgx-lkl-debug/bin/../lib/libsgxlkl.so.signed
[   SGX-LKL  ] oe_create_enclave...
[   SGX-LKL  ] Enclave config: {"args":["pthread_join-test"],"clock_res":[{"resolution":"0000000000000001"},{"resolution":"0000000000000001"},{"resolution":"0000000000000000"},{"resolution":"0000000000000000"},{"resolution":"0000000000000001"},{"resolution":"00000000003d0900"},{"resolution":"00000000003d0900"},{"resolution":"0000000000000001"}],"cwd":"/","env":[],"esleep":"16000","espins":"500","ethreads":"8","exit_status":"full","format_version":"1","fsgsbase":true,"host_import_env":[],"hostname":"lkl","hostnet":false,"image_sizes":{"num_heap_pages":"262144","num_stack_pages":"1024"},"io":{"block":true,"console":true,"network":true},"kernel_cmd":"mem=32M","kernel_verbose":true,"max_user_threads":"256","mmap_files":"shared","mode":"hw_debug","mounts":[],"net_gw4":"10.0.1.254","net_ip4":"10.0.1.1","net_mask4":"24","oe_heap_pagecount":"8192","root":{"key":"","key_id":null,"overlay":false,"readonly":false,"roothash":null,"roothash_offset":"0"},"stacksize":"524288","swiotlb":true,"sysctl":null,"tap_mtu":"0","verbose":true,"wg":{"ip":"10.0.2.1","key":null,"listen_port":"56002","peers":[]}}
result=0 (OE_OK)
[   SGX-LKL  ] sgxlkl_enclave_init(ethread_id=0)
[[  SGX-LKL ]] sgxlkl_enclave_init(): enter
[[  SGX-LKL ]] _sgxlkl_enclave_show_attribute(): enclave base=0x7fd000000000 size=64.0 GiB
[[  SGX-LKL ]] _sgxlkl_enclave_show_attribute(): enclave heap base=0x7fd000b03000 size=1024.0 MiB end=0x7fd040b03000
[[  SGX-LKL ]] sgxlkl_enclave_init(): calling _dlstart_c()
[[  SGX-LKL ]] _register_enclave_signal_handlers(): Registering OE exception handler...
[[  SGX-LKL ]] __libc_init_enclave(): calling enclave_mman_init()
[[  SGX-LKL ]] __libc_init_enclave(): calling _lthread_sched_init()
[[  SGX-LKL ]] lkl_start_init(): calling register_lkl_syscall_overrides()
[[  SGX-LKL ]] lkl_start_init(): calling initialize_enclave_event_channel()
[[  SGX-LKL ]] lkl_start_init(): kernel command line: 'mem=32M console=hvc0'
[[  SGX-LKL ]] lkl_start_init(): Disk 0: Disk encryption: no
[[  SGX-LKL ]] lkl_start_init(): Disk 0: Disk is writable: yes
[[  SGX-LKL ]] lkl_start_init(): lkl_start_kernel() called
[    0.000000] Linux version 5.3.0+ (AzDevOps@cisca24ba0002X9) (gcc version 7.5.0 (Ubuntu 7.5.0-3ubuntu1~18.04)) #1 Fri Aug 7 17:18:52 UTC 2020
[    0.000000] memblock address range: 0x7fd03e882000 - 0x7fd040882000
[    0.000000] Built 1 zonelists, mobility grouping on.  Total pages: 8080
[    0.000000] Kernel command line: mem=32M console=hvc0 virtio_mmio.device=256@0x1000000:1
[    0.000000] Dentry cache hash table entries: 4096 (order: 3, 32768 bytes, linear)
[    0.000000] Inode-cache hash table entries: 2048 (order: 2, 16384 bytes, linear)
[    0.000000] mem auto-init: stack:off, heap alloc:off, heap free:off
[    0.000000] Memory available: 32176k/32768k RAM
[    0.000000] SLUB: HWalign=32, Order=0-3, MinObjects=0, CPUs=1, Nodes=1
[    0.000000] NR_IRQS: 4096
[    0.000000] lkl: irqs initialized
[    0.000000] clocksource: lkl: mask: 0xffffffffffffffff max_cycles: 0x1cd42e4dffb, max_idle_ns: 881590591483 ns
[    0.000000] lkl: time and timers initialized (irq2)
[    0.000003] printk: console [hvc0] enabled
[    0.000003] pid_max: default: 4096 minimum: 301
[    0.000003] Mount-cache hash table entries: 512 (order: 0, 4096 bytes, linear)
[    0.000004] Mountpoint-cache hash table entries: 512 (order: 0, 4096 bytes, linear)
[    0.000015] devtmpfs: initialized
[    0.014980] random: get_random_bytes called from 0x00007fd0006319de with crng_init=0
[    0.014981] clocksource: jiffies: mask: 0xffffffff max_cycles: 0xffffffff, max_idle_ns: 19112604462750000 ns
[    0.015487] futex hash table entries: 16 (order: -4, 384 bytes, linear)
[    0.015487] xor: automatically using best checksumming function   8regs     
[    0.015489] NET: Registered protocol family 16
[    0.018930] clocksource: Switched to clocksource lkl
[    0.018932] NET: Registered protocol family 2
[    0.019434] tcp_listen_portaddr_hash hash table entries: 256 (order: 0, 4096 bytes, linear)
[    0.019434] TCP established hash table entries: 512 (order: 0, 4096 bytes, linear)
[    0.019434] TCP bind hash table entries: 512 (order: 0, 4096 bytes, linear)
[    0.019434] TCP: Hash tables configured (established 512 bind 512)
[    0.019937] UDP hash table entries: 128 (order: 0, 4096 bytes, linear)
[    0.019938] UDP-Lite hash table entries: 128 (order: 0, 4096 bytes, linear)
[    0.019940] NET: Registered protocol family 1
[    0.019940] virtio-mmio: Registering device virtio-mmio.0 at 0x1000000-0x10000ff, IRQ 1.
[    0.020443] workingset: timestamp_bits=62 max_order=14 bucket_order=0
[    0.021503] NET: Registered protocol family 38
[    0.021503] io scheduler mq-deadline registered
[    0.021504] io scheduler kyber registered
[    0.116550] software IO TLB: mapped [mem 0x7fe2a5d69000-0x7fe2a9d69000] (64MB)
[    0.120132] loop: module loaded
[    0.128338] device-mapper: ioctl: 4.40.0-ioctl (2019-01-18) initialised: dm-devel@redhat.com
[    0.138323] wireguard: WireGuard 0.0.20191219 loaded. See www.wireguard.com for information.
[    0.138879] wireguard: Copyright (C) 2015-2019 Jason A. Donenfeld <Jason@zx2c4.com>. All Rights Reserved.
[    0.139384] NET: Registered protocol family 10
[    0.139387] Segment Routing with IPv6
[    0.139388] sit: IPv6, IPv4 and MPLS over IPv4 tunneling driver
[    0.141403] This architecture does not have kernel memory protection.
[    0.141403] Run /init as init process
[[  SGX-LKL ]] lkl_start_init(): lkl_start_kernel() finished
[[  SGX-LKL ]] lkl_start_init(): creating LKL termination thread
[[  SGX-LKL ]] create_lkl_termination_thread(): enter
[[  SGX-LKL ]] lkl_start_init(): calling lkl_mount_virtual()
[[  SGX-LKL ]] lkl_termination_thread(): enter
[[  SGX-LKL ]] lkl_termination_thread(): Performed LKL syscall to get host task allocated (pid=36)
[[  SGX-LKL ]] lkl_start_init(): calling init_random()
[[  SGX-LKL ]] init_random(): Adding entropy to entropy pool
[    0.143472] random: crng init done
[[  SGX-LKL ]] init_enclave_clock(): Setting enclave realtime clock
[[  SGX-LKL ]] wg0 has public key 09RO378oOw/vQVjwoMrIsao7lxjWvX+MYScNrz1qHwY=
[    0.147519] virtio_blk virtio1: [vda] 10240 512-byte logical blocks (5.24 MB/5.00 MiB)
[[  SGX-LKL ]] lkl_mount_disk(): lkl_mount_disk(dev="/dev/vda", mnt="/mnt/vda", ro=0)
[    0.150256] EXT4-fs (vda): mounted filesystem without journal. Opts: 
[[  SGX-LKL ]] lkl_mount_disks(): Set working directory: /
[[  SGX-LKL ]] libc_start_main_stage2(): Calling app main: pthread_join-test
Creating worker thread (run=0)
Joining worker thread...
Thread joined (ret=0, run=0)
Creating worker thread (run=1000)
Joining worker thread...
Thread joined (ret=0, run=1000)
Makefile:29: recipe for target 'run-hw' failed
make: *** [run-hw] Terminated
make run-hw: TIMED OUT after 60 secs
Test run completed with EXIT CODE 124
```

--- a/src/sched/lthread.c
+++ b/src/sched/lthread.c
@@ -46,8 +46,8 @@
 #include "enclave/lthread_int.h"
 #include "enclave/sgxlkl_t.h"
 #include "enclave/ticketlock.h"
-#include "shared/tree.h"
 #include "openenclave/corelibc/oemalloc.h"
+#include "shared/tree.h"
 
 #include "openenclave/corelibc/oemalloc.h"
 #include "openenclave/corelibc/oestring.h"
@@ -191,9 +191,7 @@ void __schedqueue_inc()
     a_inc(&schedqueuelen);
 }
 
-void lthread_sched_global_init(
-    size_t sleepspins_,
-    size_t sleeptime_ns_)
+void lthread_sched_global_init(size_t sleepspins_, size_t sleeptime_ns_)
 {
     sleepspins = sleepspins_;
     sleeptime_ns = sleeptime_ns_;
@@ -204,7 +202,7 @@ void lthread_sched_global_init(
 void lthread_notify_completion(void)
 {
     SGXLKL_TRACE_THREAD(
-        "[tid=%-3d] lthread_notify_completion. \n", lthread_self()->tid);
+        "[%4d] lthread_notify_completion\n", lthread_self()->tid);
     _lthread_should_stop = true;
 }
 
@@ -243,7 +241,7 @@ void lthread_run(void)
                 pauses = sleepspins;
                 a_dec(&schedqueuelen);
                 SGXLKL_TRACE_THREAD(
-                    "[tid=%-3d] lthread_run(): lthread_resume (dequeue)\n",
+                    "[%4d] lthread_run(): lthread_resume (dequeue)\n",
                     lt ? lt->tid : -1);
                 _lthread_resume(lt);
             }
@@ -281,7 +279,7 @@ void lthread_run(void)
         if (_lthread_should_stop)
         {
             SGXLKL_TRACE_THREAD(
-                "[tid=%-3d] lthread_run(): quitting\n", lt ? lt->tid : -1);
+                "[%4d] lthread_run(): quitting\n", lt ? lt->tid : -1);
             break;
         }
     }
@@ -296,7 +294,7 @@ void _lthread_desched_sleep(struct lthread* lt)
 {
     ticket_lock(&sleeplock);
     SGXLKL_TRACE_THREAD(
-        "[tid=%-3d] _lthread_desched_sleep() TICKET_LOCK lock=SLEEPLOCK tid=%d "
+        "[%4d] _lthread_desched_sleep() TICKET_LOCK lock=SLEEPLOCK tid=%d "
         "\n",
         (lthread_self() ? lthread_self()->tid : 0),
         lt->tid);
@@ -312,7 +310,7 @@ void _lthread_desched_sleep(struct lthread* lt)
     ticket_unlock(&sleeplock);
 
     SGXLKL_TRACE_THREAD(
-        "[tid=%-3d] _lthread_desched_sleep() TICKET_UNLOCK lock=SLEEPLOCK "
+        "[%4d] _lthread_desched_sleep() TICKET_UNLOCK lock=SLEEPLOCK "
         "tid=%d\n",
         (lthread_self() ? lthread_self()->tid : 0),
         lt->tid);
@@ -336,6 +334,10 @@ static void _lthread_lock(struct lthread* lt)
 static void _lthread_unlock(struct lthread* lt)
 {
     a_barrier();
+
+    // We should never unlock an lthread that is not locked.
+    SGXLKL_ASSERT(lt->attr.state & BIT(LT_ST_BUSY));
+
     lt->attr.state &= CLEARBIT(LT_ST_BUSY);
 }
 
@@ -378,7 +380,8 @@ void _lthread_free(struct lthread* lt)
         lt->robust_list.head = *rp;
         int cont = a_swap(&m->_m_lock, lt->tid | 0x40000000);
         lt->robust_list.pending = 0;
-        if (cont < 0 || waiters) {
+        if (cont < 0 || waiters)
+        {
             enclave_futex_wake((int*)&m->_m_lock, 1);
         }
     }
@@ -629,12 +632,13 @@ int lthread_create_primitive(
     lt->itlssz = libc.tls_size;
     if (libc.tls_size)
     {
-        if ((intptr_t)(lt->itls = (uint8_t*)enclave_mmap(
-                 0,
-                 lt->itlssz,
-                 0, /* map_fixed */
-                 PROT_READ | PROT_WRITE,
-                 1 /* zero_pages */)) < 0)
+        if ((intptr_t)(
+                lt->itls = (uint8_t*)enclave_mmap(
+                    0,
+                    lt->itlssz,
+                    0, /* map_fixed */
+                    PROT_READ | PROT_WRITE,
+                    1 /* zero_pages */)) < 0)
         {
             lthread_dealloc(lt);
             return -1;
@@ -666,8 +670,7 @@ int lthread_create_primitive(
 
     a_inc(&libc.threads_minus_1);
 
-    SGXLKL_TRACE_THREAD(
-        "[tid=%-3d] create: thread_count=%d\n", lt->tid, thread_count);
+    SGXLKL_TRACE_THREAD("[%4d] create: count=%d\n", lt->tid, thread_count);
 
 #if DEBUG
     struct lthread_queue* new_ltq =
@@ -723,12 +726,13 @@ int lthread_create(
         return -1;
     }
     lt->attr.stack = attrp ? attrp->stack : 0;
-    if ((!lt->attr.stack) && ((intptr_t)(lt->attr.stack = enclave_mmap(
-                                   0,
-                                   stack_size,
-                                   0, /* map_fixed */
-                                   PROT_READ | PROT_WRITE,
-                                   1 /* zero_pages */)) < 0))
+    if ((!lt->attr.stack) && ((intptr_t)(
+                                  lt->attr.stack = enclave_mmap(
+                                      0,
+                                      stack_size,
+                                      0, /* map_fixed */
+                                      PROT_READ | PROT_WRITE,
+                                      1 /* zero_pages */)) < 0))
     {
         lthread_dealloc(lt);
         return -1;
@@ -739,12 +743,13 @@ int lthread_create(
     lt->itlssz = libc.tls_size;
     if (libc.tls_size)
     {
-        if ((intptr_t)(lt->itls = (uint8_t*)enclave_mmap(
-                 0,
-                 lt->itlssz,
-                 0, /* map_fixed */
-                 PROT_READ | PROT_WRITE,
-                 1 /* zero_pages */)) < 0)
+        if ((intptr_t)(
+                lt->itls = (uint8_t*)enclave_mmap(
+                    0,
+                    lt->itlssz,
+                    0, /* map_fixed */
+                    PROT_READ | PROT_WRITE,
+                    1 /* zero_pages */)) < 0)
         {
             lthread_dealloc(lt);
             return -1;
@@ -778,8 +783,7 @@ int lthread_create(
 
     a_inc(&libc.threads_minus_1);
 
-    SGXLKL_TRACE_THREAD(
-        "[tid=%-3d] create: thread_count=%d\n", lt->tid, thread_count);
+    SGXLKL_TRACE_THREAD("[%4d] create: count=%d\n", lt->tid, thread_count);
 
 #if DEBUG
     struct lthread_queue* new_ltq =
@@ -835,8 +839,7 @@ void lthread_exit(void* ptr)
     /* switch thread to exiting state */
     _lthread_lock(lt);
 
-    SGXLKL_TRACE_THREAD(
-        "[tid=%-3d] thread_exit: thread_count=%d\n", lt->tid, thread_count);
+    SGXLKL_TRACE_THREAD("[%4d] thread_exit: count=%d\n", lt->tid, thread_count);
 
     lt->yield_cbarg = ptr;
     lt->attr.state |= BIT(LT_ST_EXITED);
@@ -857,36 +860,44 @@ int lthread_join(struct lthread* lt, void** ptr, uint64_t timeout)
     struct lthread* current = lthread_get_sched()->current_lthread;
     if (lt->attr.state & BIT(LT_ST_DETACH))
     {
+        SGXLKL_TRACE_THREAD(
+            "[%4d] join (detached): tid=%d count=%d\n",
+            (lthread_self() ? lthread_self()->tid : 0),
+            lt->tid,
+            thread_count);
         return EINVAL;
     }
     _lthread_lock(lt);
     if (lt->attr.state & BIT(LT_ST_EXITED))
     {
         SGXLKL_TRACE_THREAD(
-            "[tid=%-3d] join:  tid=%d count=%d\n",
+            "[%4d] join (exited): tid=%d count=%d\n",
             (lthread_self() ? lthread_self()->tid : 0),
             lt->tid,
             thread_count);
-
-        /* we can test for exited flag only with lock acquired */
-        _lthread_unlock(lt);
     }
     else
     {
         SGXLKL_TRACE_THREAD(
-            "[tid=%-3d] join:  tid=%d count=%d\n",
+            "[%4d] join (waiting): tid=%d count=%d\n",
             (lthread_self() ? lthread_self()->tid : 0),
             lt->tid,
             thread_count);
 
         /* thread is still running, set current lthread as joiner */
-        if (a_cas_p(&lt->lt_join, 0, current) != 0)
+        void* null_ptr = NULL;
+        if (!atomic_compare_exchange_strong(&lt->lt_join, &null_ptr, current) !=
+            0)
         {
             /* there already is a joiner */
             _lthread_unlock(lt);
             return EINVAL;
         }
         _lthread_yield_cb(current, (void*)_lthread_unlock, lt);
+
+        // Reacquire the lthread lock before we start freeing the lthread. It may
+        // still be exiting concurrently.
+        _lthread_lock(lt);
     }
     if (ptr)
     {
@@ -974,7 +985,7 @@ int lthread_setcancelstate(int new, int* old)
  * accessed.  `lthread_current()` is always safe to use here as is any lthread
  * that has not yet been scheduled.
  */
-static struct lthread_tls* lthread_findtlsslot(struct lthread *lt, long key)
+static struct lthread_tls* lthread_findtlsslot(struct lthread* lt, long key)
 {
     struct lthread_tls *d, *d_tmp;
     LIST_FOREACH_SAFE(d, &lt->tls, tls_next, d_tmp)
@@ -1108,7 +1119,8 @@ void lthread_dump_all_threads(void)
 {
     struct lthread_queue* lt_queue = __active_lthreads;
 
-    sgxlkl_info("=============================================================\n");
+    sgxlkl_info(
+        "=============================================================\n");
     sgxlkl_info("Stack traces for all lthreads:\n");
 
     for (int i = 1; lt_queue; i++)
@@ -1117,13 +1129,21 @@ void lthread_dump_all_threads(void)
         SGXLKL_ASSERT(lt);
         int tid = lt->tid;
         char* funcname = lt->funcname;
-        sgxlkl_info("-------------------------------------------------------------\n");
-        sgxlkl_info("%s%i: tid=%i [%s]\n", lt == lthread_self() ? "*" : "", i, tid, funcname);
-        sgxlkl_print_backtrace(lt == lthread_self() ? __builtin_frame_address(0) : lt->ctx.ebp);
+        sgxlkl_info(
+            "-------------------------------------------------------------\n");
+        sgxlkl_info(
+            "%s%i: tid=%i [%s]\n",
+            lt == lthread_self() ? "*" : "",
+            i,
+            tid,
+            funcname);
+        sgxlkl_print_backtrace(
+            lt == lthread_self() ? __builtin_frame_address(0) : lt->ctx.ebp);
 
         lt_queue = lt_queue->next;
     }
 
-    sgxlkl_info("=============================================================\n");
+    sgxlkl_info(
+        "=============================================================\n");
 }
 #endif

